salt/states/boto_s3.py

```diff
@@ -0,0 +1,310 @@
+# -*- coding: utf-8 -*-
+'''
+Manage S3 Resources
+=================
+
+.. versionadded:: 2016.3.0
+
+Manage S3 resources. Be aware that this interacts with Amazon's services,
+and so may incur charges.
+
+This module uses ``boto3``, which can be installed via package, or pip.
+
+This module accepts explicit AWS credentials but can also utilize
+IAM roles assigned to the instance through Instance Profiles. Dynamic
+credentials are then automatically obtained from AWS API and no further
+configuration is necessary. More information available `here
+<http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html>`_.
+
+If IAM roles are not used you need to specify them either in a pillar file or
+in the minion's config file:
+
+.. code-block:: yaml
+
+    s3.keyid: GKTADJGHEIQSXMKKRBJ08H
+    s3.key: askdjghsdfjkghWupUjasdflkdfklgjsdfjajkghs
+
+It's also possible to specify ``key``, ``keyid`` and ``region`` via a profile,
+either passed in as a dict, or as a string to pull from pillars or minion
+config:
+
+.. code-block:: yaml
+
+    myprofile:
+        keyid: GKTADJGHEIQSXMKKRBJ08H
+        key: askdjghsdfjkghWupUjasdflkdfklgjsdfjajkghs
+            region: us-east-1
+
+.. code-block:: yaml
+
+    Ensure s3 object exists:
+        boto_s3.object_present:
+            - name: s3-bucket/s3-key
+            - source: /path/to/local/file
+            - region: us-east-1
+            - keyid: GKTADJGHEIQSXMKKRBJ08H
+            - key: askdjghsdfjkghWupUjasdflkdfklgjsdfjajkghs
+            - profile: my-profile
+
+:depends: boto3
+'''
+
+# Import Python Libs
+from __future__ import absolute_import
+import copy
+import difflib
+import logging
+
+import yaml
+
+# Import Salt libs
+import salt.ext.six as six
+import salt.utils
+
+log = logging.getLogger(__name__)
+
+
+def __virtual__():
+    '''
+    Only load if boto is available.
+    '''
+    if 'boto_s3.get_object_metadata' not in __salt__:
+        return False
+    return 'boto_s3'
+
+
+# Keys for `extra_args` that we support.
```

Comment:
If you use the approach I mention above, these arrays become unneeded, as AWS itself will happily fact check all your args and notify if any are malformed, or invalid.  You can still do any merging needed of "extra_args_from_pillar" into the passed in parameters, as you do below.  As a benefit, you'll find that the currently unsupported options you mention magically start working, as long as they're correctly specified by the caller (e.g. the pillar source).

Reply:
I did originally just pass through the args directly, but added this validation after hitting some issues. Unfortunately, not all of the arguments are handled symmetrically between what we send to S3 and what S3 stores/returns. For one, we'll need to send different subsets of args to different S3 APIs (see `UPLOAD_ONLY_ARGS` and `GET_METADATA_EXTRA_ARGS`). Second, some args that are valid for upload aren't returned (`RequestPayer` vs `RequestCharged`), so we need to exclude those from the diff. Additionally, the various `Grant*` options on upload act as shortcuts for various ACL settings, and are not returned as-is but in a converted form in the ACL data.

In short, we can't just pass all args to all the S3 endpoints we hit and rely on AWS validation, and we also need to add custom code that handles conversions properly for the purposes of diffing and comparing changes. It's important to me that we don't spuriously report changes, because it causes us to make unnecessary API calls, hit ratelimits and chew bandwidth.

Reply:
Understood, I had to do something similar in the boto3_elasticache module.  MOST of the params mapped directly, but some were renamed while others returned different formats (on describe) than what had been sent (by create or modify).  Some were ONLY valid for modify operations, not creates, etc, etc  It's annoying for sure.

Even so, there were a large number of args, the vast majority of which DID map cleanly and easily, so I took the approach of simply munging the few that needed "touching" into the common format for comparisons to work right.  But that's certainly an implementation choice which can fall either way :)

If you want a view on one way to manage this sort of "munging" where needed but otherwise letting things pass through where possible, take a look at {modules,states}/boto3_elasticache.py, particularly`_diff_cache_cluster()`.

I absolutely understand the desire to avoid unneeded change records being sent - we're not nearly at your scale and we already get plenty of throttles from the API as it is...

TL;DR  I'm fine with the implementation as it stands, and will happily approve the change set, other than the one following point, which I hope you'll take to heart and address if you can.

There is a huge forward-looking benefit to keeping camelcase args in that format, rather than mapping them to arbitrary snake_case params.  To be honest, the latter is just buying yourself a lifetime of mapping woes, most of which (not all as you point out but the vast majority) would be avoided if you leveraged the boto3 parameter names (and function names as mentioned above) wherever possible.  The reason snake_case was the norm in the original boto2 based modules was because boto2 ITSELF had mapped all the camelcase stuff to snake, and thus snake_case was the way to directly map salt params to boto params.

This is no longer the case - boto3 uses CamelCase, and creating lots of unneeded code-lines simply to convert the one to the other because it looks familiar from writing boto2 code is, as mentioned, just buying future you (and future me and every other developer touching it) a lot of unnecessary grief and LOC.

I know it's a pain when the code is working already as-is, but throwing away all my other input and feedback, the strongest suggestion I can make for this and future boto3 based code is to PLEASE use the native CamelCase NOW, so that we don't have to support complex parameter munging for EVERY parameter -- not just those small few which actually need  munging to work right -- in this module FOREVERMORE :)

Reply:
I agree that munging names is annoying. This patch actually doesn't munge any of the possible `ExtraArgs` parameters, just use the frozensets to determine which ones to pass to which S3 API calls or the diffing code. The only munging here is that the state name is used to derive the `Bucket` and `Key` args, so that the user can specify a simple `some-bucket/some/object/path` as the state name, for ergonomics and to fit Salt conventions.

Reply:
Oh, and the module injects the content hash into the metadata for diffing purposes, but I'm using the native CamelCase `Metadata` key to access that.